### ISSUE TITLE
Rule Proposal: Validate "emits" section

### ISSUE THREAD

**Issue by ilyub** [vuejs/eslint-plugin-vue#1787]
**Please describe what the rule should do:**

This rule proposal is based on this issue:
https://github.com/vuejs/core/issues/5343

It should ensure that emits section does not break vue component typing.

**What category should the rule belong to?**

Warns about a potential error (problem)

**Provide 2-3 code examples that this rule should warn about:**

Valid:
```vue
export default defineComponent({
  name: "Sample",
  emits: {
    event1(this: undefined): boolean { return true; },
    event2: (): boolean => { return true; }
  },
```

Invalid:
```vue
export default defineComponent({
  name: "Sample",
  emits: {
    event(): boolean { return true; }
  },
```

**Comment by FloEdelmann** [comment]
Makes sense. I'd suggest `vue/prefer-emits-arrow-function` as a rule name. Arrow functions work in both TypeScript and JavaScript, and the [preview docs section](https://staging.vuejs.org/guide/typescript/options-api.html#typing-component-props) linked [here](https://github.com/vuejs/core/issues/5343#issuecomment-1024899322) recommend using them. So I wouldn't bother with the first style (`event1(this: undefined) …`).

**Comment by ilyub** [comment]
> So I wouldn't bother with the first style (event1(this: undefined) …).

The first notation (event1(this: undefined)) can be useful to avoid conflict with ESLint core rule object-shorthand[https://eslint.org/docs/rules/object-shorthand]. So, I would keep it.

**Comment by FloEdelmann** [comment]
Makes sense. I think the rule should indeed be called `vue/prefer-emits-arrow-function` and work like this:

First detect if there is already type information (either only look at the function return type, or detect TypeScript usage in general somehow).

* If there isn't, provide an autofix to turn it into an arrow function. We can't be compatible with the `object-shorthand` rule then.
* If there is, don't provide an autofix, but two suggestions:
	1. Turn into arrow function
	2. Add explicit `undefined` type for `this` parameter

---

I guess that `event: function(): boolean { … }` will also be invalid?

**Comment by ilyub** [comment]
> I guess that event: function(): boolean { … } will also be invalid?

I tried to use this pattern and it also break emit typing:
![image](https://user-images.githubusercontent.com/70687432/152318736-7b86c9b4-a3c3-4e82-92ae-ea2b348f5d7a.png)
So, I guess it should also be reported.

---

But note that the pattern bellow works well (i.e. no need to report it):
```ts
  event: function(this: undefined): boolean { … }
```

**Comment by FloEdelmann** [comment]
Thanks for checking.

> But note that the pattern bellow works well (i.e. no need to report it):

Sure. That means that shorthand and longhand functions can be handled exactly the same. So the algorithm in [my comment above](https://github.com/vuejs/eslint-plugin-vue/issues/1787#issuecomment-1028769941) should be fine.
